kpatch-build/create-diff-object.c

```diff
@@ -2326,8 +2392,16 @@ static void kpatch_create_intermediate_sections(struct kpatch_elf *kelf,
 			continue;
 		if (!strcmp(sec->name, ".rela.kpatch.funcs"))
 			continue;
-		list_for_each_entry(rela, &sec->relas, list)
+		list_for_each_entry(rela, &sec->relas, list) {
 			nr++; /* upper bound on number of kpatch relas and symbols */
+			if (!strcmp(sec->name, ".rela.toc")) {
+				rela->need_dynrela =
+					(!rela->sym->sec || rela->need_dynrela);
+				continue;
+			}
+			if (may_need_dynrela(rela))
+				toc_rela(rela)->need_dynrela = 1;
+		}
```

Comment:
This is a lot better than before, but I'm still confused by this code.  Why are all the toc special cases needed in function_ptr_rela(), may_need_rela(), and here?  This code loops through all the sections, including `.rela.toc.`, so can we just do something like
```
		list_for_each_entry(rela, &sec->relas, list) {
			nr++; /* upper bound on number of kpatch relas and symbols */
			if (may_need_dynrela(rela))
				rela->need_dynrela = 1;
		}
```
along with removing all the toc_rela() calls in function_ptr_rela() and may_need_rela()?

Reply:
```
Relocation section '.rela.toc' at offset 0xcc6b0 contains 46 entries:
...
0000000000000128  0000353400000026 R_PPC64_ADDR64         0000000000000000 genl_sk_destructing_waitq + 0
0000000000000130  0000003e00000026 R_PPC64_ADDR64         0000000000000000 .text.netlink_release + 35c
0000000000000138  0000002a00000026 R_PPC64_ADDR64         0000000000000000 .text.deferred_put_nlk_sk + 8

Relocation section '.rela.text.netlink_release' at offset 0xcadf0 contains 44 entries:
...
0000000000000390  000035580000000a R_PPC64_REL24          0000000000000000 __local_bh_enable_ip + 0
0000000000000398  0000007300000032 R_PPC64_TOC16_HA       0000000000000000 .toc + 138
00000000000003a0  0000007300000040 R_PPC64_TOC16_LO_DS    0000000000000000 .toc + 138
```
When looping with:
```
        list_for_each_entry(rela, &sec->relas, list) {
			nr++; /* upper bound on number of kpatch relas and symbols */
			if (may_need_dynrela(rela))
				rela->need_dynrela = 1;
		}
```
`netlink_release + 0x3a0` is not a function symbol, but of relocation type `(R_PPC64_TOC16_HA/R_PPC64_TOC16_LO_DS)`
referring to symbol in `.rela.toc + 0x138`, We need to operate on both `rela->type` and `rela->sym->sec->rela + rela->addend`, where `netlink_release + 0x3a0` is checked for `rela-type` and  `.text.deferred_put_nlk_sk + 8` for STT_FUNC/switch label.

Without `toc_rela(netlink_release + 0x3a0)` the `.toc + 0x138` need_dynrela will not be set and while parsing `.rela.toc_rela` section it's not obvious on if a rela requires, `need_dynrela` to be set, other than for `!rela->sym->sec case`.

Alternative above loop can be re-written as:
```
        list_for_each_entry(rela, &sec->relas, list) {
            nr++; /* upper bound on number of kpatch relas and symbols */
            if (may_need_dynrela(rela))
                toc_rela(rela)->need_dynrela = 1;
        }
```
With an additional check in may_need_dynrela():
```
static int may_need_dynrela(const struct rela *rela)
{
        /* Don't over-ride need_dynrela flag if set, required for .toc sym*/
        if (rela->need_dynrela || !rela->sym->sec)
                return 1;
    ...
}
```
`may_need_dynrela()` and `function_ptr_rela()` operates on both `rela` and `rela->sym->sec->rela`. We might need to keep `toc_rela()` in both.

Reply:
Thanks for the explanation.  Sorry, I keep forgetting that the R_PPC64_TOC16_HA relocation is for the original section, not the .toc.  Can you add the above readelf snippet to a comment above the `if (may_need_dynrela(rela))` line explaining why the toc_rela() is getting need_dynrela set?  Otherwise I'll keep getting confused :-)

Can you also explain why you need to check rela->need_dynrela before setting toc_rela(rela)->need_dynrela?  I'm trying to understand how it would be possible for a ".toc + xxx) relocation to be converted to a dynrela.

Assuming we do really need that, how about doing the loop like this?

```
        list_for_each_entry(rela, &sec->relas, list) {
            nr++; /* upper bound on number of kpatch relas and symbols */
            if (!rela->need_dynrela && may_need_dynrela(rela))
                toc_rela(rela)->need_dynrela = 1;
        }
```

Also a comment for the `!strchr(toc_rela(rela)->sym->name, '.')` check in may_need_dynrela() would also be helpful.

Another thing that would make it easier for me to follow would be to rename rela1 to rela_toc in function_ptr_rela().

Thanks!

Reply:
> the if (may_need_dynrela(rela)) line explaining why the toc_rela() is getting need_dynrela set? Otherwise I'll keep getting confused :-)   
> Can you also explain why you need to check rela->need_dynrela before setting toc_rela(rela)->need_dynrela? I'm trying to understand how it would be possible for a ".toc + xxx)" relocation to be converted to a dynrela.

:) `.rela.toc` is superseded by `.rela.text.*` sections, those might have set the need_dynrela                         for entries in `.rela.toc.` Without the !rela->need_dynrela check, may_need_dynrela() for `.rela.toc` rela* will unset the need_dynrela flag, failing to rela->type check in function_ptr_rela().
`.rela.toc` rela type is always `R_PPC64_ADDR64`.

> check in may_need_dynrela() would also be helpful.

function_ptr_rela(rela) will return 0, if the rela returned by toc_rela() is NULL. Duplicate check for NULL in may_need_dynrela(), can be avoided.

Reply:
> Without the !rela->need_dynrela check, may_need_dynrela() for .rela.toc rela* will unset the need_dynrela flag, failing to rela->type check in function_ptr_rela().

Hm, but if it's already set, how would it get unset?  To clarify, I'm talking about your rewritten version of the loop:
```
        list_for_each_entry(rela, &sec->relas, list) {
            nr++; /* upper bound on number of kpatch relas and symbols */
            if (may_need_dynrela(rela))
                toc_rela(rela)->need_dynrela = 1;
        }
```

Reply:
Sorry got little mixed up here, true it will not get unset.  Having the check for need_dynrela, 
`if (!rela->need_dynrela && may_need_dynrela(rela))`  will speed the handling of `.rela.toc` entries but not sure, if its worth the optimization. Arguably `.rela.toc` is only on PPC64le and an additional check for rela entries for non-toc sections might be an overkill. I will re-spin the patch with 
```
list_for_each_entry(rela, &sec->relas, list) {
            nr++; /* upper bound on number of kpatch relas and symbols */
            if (may_need_dynrela(rela))
                toc_rela(rela)->need_dynrela = 1;
        }
```
Thanks.